packages/server/src/Hocuspocus.ts

```diff
@@ -414,15 +416,23 @@ export class Hocuspocus {
   /**
    * Create a new document by the given request
    */
-  public async createDocument(documentName: string, request: Partial<Pick<IncomingMessage, 'headers' | 'url'>>, socketId: string, connection: ConnectionConfiguration, context?: any): Promise<Document> {
-    if (this.documents.has(documentName)) {
-      const document = this.documents.get(documentName)
+  public createDocument(documentName: string, request: Partial<Pick<IncomingMessage, 'headers' | 'url'>>, socketId: string, connection: ConnectionConfiguration, context?: any): Promise<Document> {
+    if (this.loadingDocuments.has(documentName)) {
+      const documentPromise = this.loadingDocuments.get(documentName)
 
-      if (document) {
-        return document
+      if (documentPromise) {
+        return documentPromise
       }
     }
 
+    const loadDocPromise = this.loadDocument(documentName, request, socketId, connection, context)
+
+    this.loadingDocuments.set(documentName, loadDocPromise)
+
+    return loadDocPromise
+  }
```

Comment:
This map will just grow in memory and always short circuit fetching the document (since it will have the promise in memory).
I'd suggest doing a `.finally` to remove this documentName from the map afterwards to clean it up. We don't want this to act as a cache

Reply:
Good catch here on this growing infinitely!  

However, I dont quite follow using `.finally` here.  The issue that I am thinking of is that we use `loadingDocuments` as the state of whether a doc has been loaded in the form of a promise. The only time it seems acceptable to delete from this map is when we call `unloadDocument`.  I just pushed a change that removes the `Promise<Document>` when `unloadDocument` is called.

Reply:
@jordangarcia I'm unsure that your change got pushed up here, can you check that?

I think that we are thinking of this mapping a bit differently. I am thinking of this map only being useful for "re-using" a fetch to the loading document. This would make it so that two clients requesting the same document while loading would wait for the same document to finish loading.

Client A fetches Doc 1
Client A waits for Doc 1 to load
Client B fetches Doc 1
Client B waits on the same promise as Client A
Doc loads
Client A syncs
Client B syncs


Whereas if I'm understanding you correctly, you want to store the promise here and use this as the source of truth for documents that are loaded or not. When this is what the `documents` map is meant to do.

Please let me know your thoughts

Reply:
My interpretation of the code is that the `this.documents` map isn't quite intended to keep track of "loaded" documents.  Since we insert into that map before the document is loaded.  `this.documents` seems to represent the documents that this hocuspocus instance is managing, regardless of their loading state.  Changing `this.documents` to represent documents to be loaded seems like a much bigger API change since many functions rely on this mapping to understand the state of the system.
 
The `loadDocuments` map would be used as the state of the loading documents and a way for `createDocument` to guarantee a singular doc always being returned (or a promise to the resolved value).  The fact that `loadDocuments` sets the created document in `this.documents` then returns the document gaurantees referential equality between the two maps.    I do think this may seem a bit fragile to code changes that wouldn't keep these two in sync.

The key issue here is that `createDocument` was using the `this.documents` map as a cache of the documents.  To be consistent we always want `createDocument` to be `Promise<Document>` of the loaded document.  I dont really see a way around not using `loadingDocuments` as a cache here for `createDocument`.  We could add more state to the system and keep track of `Map<string, boolean>` for whether a document is loaded and then use `this.documents` instead of `this.loadingDocuments` but that just seems more complex to me.

Reply:
Yea this makes sense to me as well. `createDocument` is idempotent and will always return a Promise that resolves to the singular (referential) source of truth `Document`.

I do see how this technically is adding more "redundant" state, but I think it's quite clear and simple.

@nperez0111 Curious to hear your thoughts!

Reply:
I think this is a good solution as well. Thanks for helping out!

Would love to get this out as soon as possible.